New destination: Calliper

liqwid opened the pull request:
<!-- Hello and thank you for contributing to Segment action-destinations! -->

<!-- Before opening your pull request, make sure you have added and ran unit
     tests and tested your change locally. Refer to our testing
     documentation for more information: https://github.com/segmentio/action-destinations/blob/main/docs/testing.md -->

<!-- If you have questions or issues please open a new issue or create a new discussion
     post in Github. -->

Segment Destination implementation for [Calliper](https://calliper.app):

- Track event action
- Identify user action
- Identify group action
- Alias action
- Authentication test endpoint
- User deletion

## Testing

Added unit tests and tested the integration numerous times with our ingestion service.

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

joe-ayoub-segment commented:
Hi @liqwid , 

Apologies for the slow response. I am just back from PTO this morning. 

Thanks for raising the PR. With new Integration PRs we usually schedule a face to face call so I can get a better understanding of the use-cases for the Integration, the data being collected, which identifiers are collected, etc. 

Would you be OK scheduling a call with me so we can do this please? Please use this Calendly [link](http://calendly.com/joe_ayoub/integration-workshop). 

Best regards,
Joe

joe-ayoub-segment commented:
Hi @liqwid , 

Apologies for the slow response. I am just back from PTO this morning. 

Thanks for raising the PR. With new Integration PRs we usually schedule a face to face call so I can get a better understanding of the use-cases for the Integration, the data being collected, which identifiers are collected, etc. 

Would you be OK scheduling a call with me so we can do this please? Please use this Calendly [link](http://calendly.com/joe_ayoub/integration-workshop). 

Best regards,
Joe

joe-ayoub-segment commented:
Hi @liqwid , 

Apologies for the slow response. I am just back from PTO this morning. 

Thanks for raising the PR. With new Integration PRs we usually schedule a face to face call so I can get a better understanding of the use-cases for the Integration, the data being collected, which identifiers are collected, etc. 

Would you be OK scheduling a call with me so we can do this please? Please use this Calendly [link](http://calendly.com/joe_ayoub/integration-workshop). 

Best regards,
Joe

joe-ayoub-segment commented:
Hi @liqwid , 

Apologies for the slow response. I am just back from PTO this morning. 

Thanks for raising the PR. With new Integration PRs we usually schedule a face to face call so I can get a better understanding of the use-cases for the Integration, the data being collected, which identifiers are collected, etc. 

Would you be OK scheduling a call with me so we can do this please? Please use this Calendly [link](http://calendly.com/joe_ayoub/integration-workshop). 

Best regards,
Joe

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/index.ts:
```diff
@@ -0,0 +1,67 @@
+import alias from './alias'
+
+const destination: DestinationDefinition<Settings> = {
+  name: 'Calliper',
```
```suggestion
  name: 'Calliper Cloud (Actions)',
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/trackEvent/eventProperties.ts:
```diff
@@ -0,0 +1,235 @@
+      '@path': '$.eventId'
+    }
+  },
+  app_version: {
```
Consider using an object array field to capture all Mobile Specific data. 
Also you could do the same for web if you liked.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/trackEvent/index.ts:
```diff
@@ -0,0 +1,77 @@
+}
+
+const processData = async (request: RequestClient, settings: Settings, payload: Payload[]) => {
+  if (!settings.companyId) {
```
These lines aren't needed, as both companyId and segmentKey are required fields, and any payload without them will be dropped before perform() or performBatch() is called.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/identifyUser/index.ts:
```diff
@@ -0,0 +1,147 @@
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Identify User',
+  description: 'Set the user ID to recognise other evens by the same user',
```
Note: some Segment customers don't send identify() calls. 
Usually identify calls are used for collecting user profile data. 
Consider making the description a little more clear for the customer. i.e what happens when an identify() call is mapped to Calliper?

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/identifyUser/index.ts:
```diff
@@ -0,0 +1,147 @@
+      type: 'string',
+      required: false,
+      default: {
+        '@path': '$.traits.firstName'
```
```suggestion
        '@path': '$.traits.first_name'
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/identifyUser/index.ts:
```diff
@@ -0,0 +1,147 @@
+      type: 'string',
+      required: false,
+      default: {
+        '@path': '$.traits.lastName'
```
```suggestion
        '@path': '$.traits.last_name'
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/identifyUser/index.ts:
```diff
@@ -0,0 +1,147 @@
+      label: 'Email',
+      type: 'string',
+      required: false,
+      default: {
+        '@if': {
+          exists: { '@path': '$.email' },
+          then: { '@path': '$.email' },
+          else: { '@path': '$.traits.email' }
+        }
+      }
```
```suggestion
      default: {
        '@path': '$.traits.email'
      }
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/identifyUser/index.ts:
```diff
@@ -0,0 +1,147 @@
+      }
+    },
+    created_at: {
+      description: 'The time the user signed up to your system',
```
Please provide a better description

```suggestion
      description: 'The timestamp for when the user profile was first created',
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/identifyUser/index.ts:
```diff
@@ -0,0 +1,147 @@
+    }
+  },
+  perform: async (request, { payload, settings }) => {
+    if (!settings.companyId) {
+      throw new IntegrationError('Missing company id', 'Missing required field', 400)
+    }
+    if (!settings.segmentKey) {
+      throw new IntegrationError('Missing segment key', 'Missing required field', 400)
+    }
```
Not needed

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/identifyUser/index.ts:
```diff
@@ -0,0 +1,147 @@
+          last_name: payload.last_name,
+          name,
+          phone: payload.phone,
+          traits: omit(payload.traits, ['created', 'email', 'firstName', 'lastName', 'name', 'username', 'phone'])
```
```suggestion
          traits: omit(payload.traits, ['created', 'email', 'firstName', 'first_name', 'lastName', last_name', 'name', 'username', 'phone'])
```

joe-ayoub-segment reviewed (changes requested):
Generally very good! 
Few minor recommendations to be looked at. please see inline. 

Also please add some presets.

liqwid commented:
@joe-ayoub-segment thanks for review and all of the advices, I've addressed the comments

decided not to combine mobile/web fields into a single one for now since I didn't find a non-array field example of setting defaults; the way it is right now should do for the time being

joe-ayoub-segment commented:
Thanks for updating the PR @liqwid . I've scheduled time to review tomorrow. 
Best regards,
Joe

joe-ayoub-segment commented:
Hi @liqwid could you provide me access to your fork please?
I need to make a couple of minor changes.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/calliper/trackEvent/eventProperties.ts:
```diff
@@ -0,0 +1,231 @@
+    description:
+      'A random id that is unique to an event. ID is being used to prevent event duplication. All the events that share the same unique id besides the first one will be ignored.',
+    default: {
+      '@path': '$.eventId'
```
```suggestion
      '@path': '$.messageId'
```

joe-ayoub-segment reviewed (changes requested):
Actually, don't bother giving me access to your fork. 
Only 1 minor thing to change - I'd recommend mapping your EventID field to $.messageId.

liqwid commented:
@joe-ayoub-segment updated, thanks!

liqwid commented:
@joe-ayoub-segment seems like the snyk check fails because of auth, is it something we can address on our side?

joe-ayoub-segment reviewed (approved):
Looks good. 
Reviewed with Partner. Partner applied recommendations, etc.

joe-ayoub-segment commented:
hi @liqwid this PR has been deployed. 
I'll email you with next steps.
